src/install/dracut-install.c

```diff
@@ -1367,81 +1367,95 @@ static int install_all(int argc, char **argv)
                 return ret;
         }
+
         kmod_list_foreach(l, list) {
                 const char *key = kmod_module_info_get_key(l);
-                const char *value = NULL;
+                const char *fwname = NULL;
+                bool found, is_glob;
 
                 if (!streq("firmware", key))
                         continue;
 
-                value = kmod_module_info_get_value(l);
-                log_debug("Firmware %s", value);
-                ret = -1;
-                STRV_FOREACH(q, firmwaredirs) {
-                        _cleanup_free_ char *fwpath = NULL;
-                        struct stat sb;
+                fwname = kmod_module_info_get_value(l);
+                log_debug("Firmware %s", fwname);
 
-                        _asprintf(&fwpath, "%s/%s", *q, value);
-
-                        if ((strstr(value, "*") != 0 || strstr(value, "?") != 0 || strstr(value, "[") != 0)
-                            && stat(fwpath, &sb) != 0) {
-                                size_t i;
-                                _cleanup_globfree_ glob_t globbuf;
+                found = false;
+                is_glob = strstr(fwname, "*") || strstr(fwname, "?") || strstr(fwname, "[");
 
-                                glob(fwpath, 0, NULL, &globbuf);
-                                for (i = 0; i < globbuf.gl_pathc; i++) {
-                                        ret = install_firmware_fullpath(globbuf.gl_pathv[i]);
-                                        if (ret != 0) {
-                                                log_info("Possible missing firmware %s for kernel module %s", value,
-                                                         kmod_module_get_name(mod));
-                                        }
-                                }
-                        } else {
-                                ret = install_firmware_fullpath(fwpath);
-                                if (ret != 0) {
-                                        log_info("Possible missing firmware %s for kernel module %s", value,
-                                                 kmod_module_get_name(mod));
-                                }
+                STRV_FOREACH(fwdir, firmwaredirs) {
+                        if (install_firmware_fullpath(*fwdir, fwname, "") == 0) {
+                                found = true;
+                                if (!is_glob)
+                                        continue;
+                        }
+                        if (install_firmware_fullpath(*fwdir, fwname, ".zst") == 0) {
+                                found = true;
+                                if (!is_glob)
+                                        continue;
+                        }
+                        if (install_firmware_fullpath(*fwdir, fwname, ".xz") == 0) {
+                                found = true;
+                                if (!is_glob)
+                                        continue;
```

Comment:
The approach looks sound; I think it may make more sense to roll this loop up tho:
```c
static const char * const * firmware_suffixes[] = {"", ".zst", ".xz", NULL};
// ...
                STRV_FOREACH(fwdir, firmwaredirs) {
                        STRV_FOREACH(suff, firmware_suffixes) {
                                if (install_firmware_fullpath(*fwdir, fwname, suff) == 0) {
                                        found = true;
                                        if (!is_glob)
                                                continue;
                                }
                        }
                }
```

Reply:
Can do that. But maybe is a good moment to ask: do we need to install same firmware from different folders? i.e. from /lib/firmware/updates and also from /lib/firmware. Currently that's what is done, and I maintained it, but is it useful?

Reply:
well, we need to correctly mirror the kernel's loading pattern – drivers/base/firmware_loader/main.c – the kernel searches for firmware A in all paths, then for firmware A.zst in all paths, then for firmware A.xz in all paths

if you invert the loop stacking to be `for(suff : suffixes) for(path : paths) if(install(path, firmware, suff) == ok) break 2;` then we could theoretically early-exit